File: cads-modern/src/utils/featureDetection.js
/**
 * Feature detection using OpenCV.js ORB (Oriented FAST and Rotated BRIEF)
 * This is the same algorithm used in the original CADS
 */

export async function detectFeatures(imagePath) {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.crossOrigin = 'anonymous';

    img.onload = () => {
      try {
        const cv = window.cv;

        if (!cv || !cv.imread) {
          reject(new Error('OpenCV is not fully initialized.'));
          return;
        }

        // Create canvas and load image
        const canvas = document.createElement('canvas');
        canvas.width = img.width;
        canvas.height = img.height;
        const ctx = canvas.getContext('2d');
        ctx.drawImage(img, 0, 0);

        // Convert to OpenCV Mat
        let src = cv.imread(canvas);

        // Convert to grayscale
        let gray = new cv.Mat();
        cv.cvtColor(src, gray, cv.COLOR_RGBA2GRAY);

        // Apply Gaussian blur to reduce noise
        let blurred = new cv.Mat();
        cv.GaussianBlur(gray, blurred, new cv.Size(5, 5), 0);

        // Create ORB detector (same as original CADS)
        const orb = new cv.ORB(1000); // Detect up to 1000 features

        // Detect keypoints and compute descriptors
        let keypoints = new cv.KeyPointVector();
        let descriptors = new cv.Mat();
        let mask = new cv.Mat();

        orb.detectAndCompute(blurred, mask, keypoints, descriptors);

        // Copy descriptors out as a compact byte array (rows x cols, CV_8U).
        // Much smaller in memory than nested JS arrays — matters for
        // thousand-image studies.
        const descRows = descriptors.rows;
        const descCols = descriptors.cols || 32;
        const descriptorData = new Uint8Array(
          descriptors.data.subarray(0, descRows * descCols)
        );

        const keypointCount = keypoints.size();

        // Clean up (everything allocated by OpenCV must be freed manually)
        src.delete();
        gray.delete();
        blurred.delete();
        mask.delete();
        keypoints.delete();
        descriptors.delete();
        orb.delete();

        resolve({
          descriptors: descriptorData,
          descRows,
          descCols,
          keypointCount
        });
      } catch (error) {
        reject(error);
      }
    };

    img.onerror = () => {
      reject(new Error(`Failed to load image: ${imagePath}`));
    };

    // In Electron with webSecurity: false, we can load local files directly
    const fileUrl = imagePath.startsWith('file://') ? imagePath : `file://${imagePath}`;
    img.src = fileUrl;
  });
}

/**
 * Build a reusable cv.Mat from stored descriptor bytes. Building these once
 * per image (instead of once per pair) makes large studies far faster.
 */
export function buildDescriptorMat(feature) {
  const cv = window.cv;
  const mat = new cv.Mat(feature.descRows, feature.descCols || 32, cv.CV_8U);
  if (feature.descRows > 0) {
    mat.data.set(feature.descriptors);
  }
  return mat;
}

/**
 * Match two descriptor Mats with a shared BFMatcher.
 * Returns a distance score (lower = more similar).
 */
export function matchDescriptorMats(bf, desc1, desc2) {
  const cv = window.cv;

  if (desc1.rows === 0 || desc2.rows === 0) {
    return 1000; // Maximum distance if no features
  }

  const matches = new cv.DMatchVector();
  try {
    bf.match(desc1, desc2, matches);

    if (matches.size() === 0) {
      return 1000;
    }

    // Calculate average distance of matches
    let totalDistance = 0;
    for (let i = 0; i < matches.size(); i++) {
      totalDistance += matches.get(i).distance;
    }
    const avgDistance = totalDistance / matches.size();

    // Normalize by number of matches (fewer matches = less similar)
    const matchRatio = matches.size() / Math.max(desc1.rows, desc2.rows);
    return avgDistance / (matchRatio + 0.01);
  } catch (error) {
    console.error('Error matching features:', error);
    return 1000;
  } finally {
    matches.delete();
  }
}

File: cads-modern/python/feature_detection.py
"""
Feature detection using OpenCV ORB (Oriented FAST and Rotated BRIEF)
This implements the same algorithm as the original CADS application
"""

import cv2
import numpy as np
from typing import Tuple, List, Dict, Any


class FeatureDetector:
    """Handles ORB feature detection and matching for die study images"""

    def __init__(self, n_features: int = 1000):
        """
        Initialize the ORB detector

        Args:
            n_features: Maximum number of features to detect (default: 1000)
        """
        self.orb = cv2.ORB_create(nfeatures=n_features)
        self.bf_matcher = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=True)

    def detect_features(self, image_path: str) -> Dict[str, Any]:
        """
        Detect ORB features in an image

        Args:
            image_path: Path to the image file

        Returns:
            Dictionary containing:
                - descriptors: Feature descriptors as list
                - keypoints: Keypoint information as list of dicts
                - success: Boolean indicating if detection succeeded
                - error: Error message if failed
        """
        try:
            # Read image
            img = cv2.imread(image_path)
            if img is None:
                return {
                    'success': False,
                    'error': f'Failed to load image: {image_path}'
                }

            # Convert to grayscale
            gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

            # Apply Gaussian blur to reduce noise (same as original CADS)
            blurred = cv2.GaussianBlur(gray, (5, 5), 0)

            # Detect keypoints and compute descriptors
            keypoints, descriptors = self.orb.detectAndCompute(blurred, None)

            if descriptors is None or len(keypoints) == 0:
                return {
                    'success': False,
                    'error': f'No features detected in image: {image_path}'
                }

            # Convert keypoints to serializable format
            keypoint_list = [
                {
                    'x': float(kp.pt[0]),
                    'y': float(kp.pt[1]),
                    'size': float(kp.size),
                    'angle': float(kp.angle),
                    'response': float(kp.response)
                }
                for kp in keypoints
            ]

            # Convert descriptors to list (numpy array -> list for JSON)
            descriptor_list = descriptors.tolist()

            return {
                'success': True,
                'descriptors': descriptor_list,
                'keypoints': keypoint_list,
                'num_features': len(keypoints)
            }

        except Exception as e:
            return {
                'success': False,
                'error': f'Error processing {image_path}: {str(e)}'
            }

    def match_features(self, desc1: List[List[int]], desc2: List[List[int]]) -> float:
        """
        Match features between two sets of descriptors using BFMatcher

        Args:
            desc1: First set of descriptors (list of lists)
            desc2: Second set of descriptors (list of lists)

        Returns:
            Distance score (lower = more similar, 0-1000 scale)
        """
        if len(desc1) == 0 or len(desc2) == 0:
            return 1000.0  # Maximum distance if no features

        try:
            # Convert lists back to numpy arrays with correct dtype
            descriptors1 = np.array(desc1, dtype=np.uint8)
            descriptors2 = np.array(desc2, dtype=np.uint8)

            # Match descriptors using BFMatcher with Hamming distance
            matches = self.bf_matcher.match(descriptors1, descriptors2)

            if len(matches) == 0:
                return 1000.0

            # Calculate average distance of matches
            total_distance = sum(match.distance for match in matches)
            avg_distance = total_distance / len(matches)

            # Normalize by match ratio (fewer matches = less similar)
            max_features = max(len(desc1), len(desc2))
            match_ratio = len(matches) / max_features

            # Calculate normalized distance
            # Lower match ratio increases the distance
            normalized_distance = avg_distance / (match_ratio + 0.01)

            return float(normalized_distance)

        except Exception as e:
            print(f'Error matching features: {e}')
            return 1000.0  # Return max distance on error


def compute_distance_matrix(features_list: List[Dict[str, Any]]) -> List[List[float]]:
    """
    Compute pairwise distance matrix for all images

    Args:
        features_list: List of feature dictionaries from detect_features()

    Returns:
        2D distance matrix as list of lists
    """
    detector = FeatureDetector()
    n = len(features_list)
    distance_matrix = [[0.0] * n for _ in range(n)]

    for i in range(n):
        for j in range(i + 1, n):
            desc1 = features_list[i]['descriptors']
            desc2 = features_list[j]['descriptors']

            distance = detector.match_features(desc1, desc2)
            distance_matrix[i][j] = distance
            distance_matrix[j][i] = distance  # Symmetric matrix

    return distance_matrix
